### ISSUE TITLE
Decision threshold tuning for minority class recall

### ISSUE THREAD

**Issue by DesmondChoy** [DesmondChoy/twinkl#20]
## Background

The current best model (run_010 CORN) achieves QWK 0.434 but has critically low minority recall (recall_minus1 = 8.9%) and hedges 82% of predictions toward the neutral class. One lever that requires **no retraining** is tuning the decision thresholds used to convert model outputs into discrete {-1, 0, +1} predictions.

### How decision thresholds affect recall

All four ordinal models produce continuous outputs (sigmoid probabilities or softmax logits) that are converted to discrete class labels via hardcoded thresholds. These thresholds implicitly encode a **uniform prior** — they assume each class is equally likely. When 85–90% of labels are neutral, this means the model needs overwhelming evidence to predict -1 or +1, resulting in systematic hedging. Lowering the threshold for non-neutral classes requires less evidence to break out of the neutral prediction, directly improving minority recall at the cost of some precision on the majority class.

### Default decision thresholds by model

| Model | Output Structure | Decoding Logic | Default Threshold |
|---|---|---|---|
| **CORAL** | 2 sigmoid cumulative probs per dim `P(Y > k)` | Count thresholds where `cum_prob > 0.5` | `0.5` on each sigmoid (`critic_ordinal.py:374`) |
| **CORN** | 2 conditional logits per dim `P(Y=k\|Y≥k)` | `corn_label_from_logits()` (coral-pytorch library) | Implicit `0.5` inside library function |
| **EMD** | 3 softmax class probs per dim | `argmax` over 3 classes | No explicit threshold — highest prob wins (`critic_ordinal.py:531`) |
| **SoftOrdinal** | 3 softmax class probs per dim | `argmax` over 3 classes | No explicit threshold — highest prob wins (`critic_ordinal.py:645`) |

### Hidden eval pathway threshold

There is an additional threshold in `eval.py:discretize_predictions()` (lines 44–46) that converts continuous MC Dropout mean predictions to classes:

```python
classes[values < -0.5] = -1
classes[values > 0.5] = 1
```

Since `predict_with_uncertainty()` averages 50 discrete `predict()` calls, the mean is continuous (e.g., -0.38), so this second threshold also gates what gets counted as -1 or +1 in evaluation. This means threshold tuning at the model level AND the eval level both matter.

---

## Three levels of threshold tuning

| Level | Where | Current Default | Tuning Approach |
|---|---|---|---|
| **1. Model decode** | `predict()` in each subclass | CORAL: `> 0.5` on sigmoid; EMD/SoftOrd: `argmax` | Lower sigmoid threshold (CORAL) or bias logits before argmax (EMD/SoftOrd) |
| **2. MC Dropout aggregation** | `predict_with_uncertainty()` mean | Mean of 50 discrete predictions, thresholded at ±0.5 | Threshold the mean more aggressively (e.g., -1 if mean < -0.3) |
| **3. Eval discretization** | `discretize_predictions()` in `eval.py` | `< -0.5` / `> 0.5` | Widen the non-neutral buckets (same effect as level 2) |

### Implementation sketch per model type

**CORAL (easiest):** Replace the hardcoded `0.5` in `predict()` with a tunable parameter. Sweep values on the validation set (e.g., 0.3–0.5 in 0.05 steps) to find the threshold that maximizes a target metric (macro-recall, QWK, or a weighted combination).

**CORN:** The `corn_label_from_logits()` function is a coral-pytorch library call with no threshold parameter. Two options: (a) reimplement the conditional-to-cumulative probability conversion and apply custom thresholds, or (b) extract the underlying probabilities and apply class-prior-aware decoding externally.

**EMD / SoftOrdinal:** Apply post-hoc logit adjustment before argmax: `logit_k -= tau * log(pi_k)`, where `pi_k` is the training class prior and `tau` is a temperature. Since neutral has ~85% prior, this shifts the decision boundary toward neutral, requiring less evidence for minority predictions. `tau` can be swept on the validation set.

**Eval pathway:** Parameterize the ±0.5 thresholds in `discretize_predictions()` and sweep jointly with model-level thresholds, or at minimum evaluate the interaction between levels 1 and 3.

### Tradeoff

Threshold tuning improves recall at the cost of precision — more true -1s detected, but also more 0s misclassified as -1. Given 82% hedging and 8.9% recall_minus1, there is substantial room to trade majority-class precision for minority-class recall before overall metrics degrade.
